src/tests/fuzzing/fuzz_pkcs11.c

```diff
@@ -0,0 +1,1076 @@
+	unsigned char        in_buffer[1024], out_buffer[1024];
+	CK_MECHANISM         mech = {0, NULL_PTR, 0};
+	size_t               out_len = 0;
+
+	/* Process options*/
+	if (set_mechanism(&data, &size, &mech) || size < 3)
+		return;
+	login_type = data[0];
+	data++; size--;
+	if (!(pin = extract_word(&data, &size)))
+		return;
+	opt_id = data;
+	opt_id_len = get_buffer(&opt_id, size, &data, &size, 256);
+
+	/* Prepare buffer for signing */
+	dec_data = data;
+	if ((dec_data_size = get_buffer(&dec_data, size, &data, &size, 1024)) == 0)
+		goto end;
+
+	/* Initialize */
+	if (fuzz_pkcs11_initialize(data, size, NULL, &session) != CKR_OK)
+		goto end;
+
+	p11->C_Login(session, login_type, (CK_UTF8CHAR *) pin, strlen(pin));
+	if (!fuzz_find_object(session, CKO_PRIVATE_KEY, &key, opt_id_len ? opt_id : NULL, opt_id_len)
+		&& !fuzz_find_object(session, CKO_SECRET_KEY, &key, opt_id_len ? opt_id : NULL, opt_id_len))
+		goto fin;
+
+	if (p11->C_DecryptInit(session, &mech, key) != CKR_OK)
+		goto fin;
+
+	p11->C_Login(session, CKU_CONTEXT_SPECIFIC, (CK_UTF8CHAR *) pin, strlen(pin));
+	out_len = sizeof(out_buffer);
+
+	memcpy(in_buffer, dec_data, dec_data_size);
+	p11->C_Decrypt(session, in_buffer, dec_data_size, out_buffer, &out_len);
+fin:
+	p11->C_CloseSession(session);
+	p11->C_Finalize(NULL);
+end:
+	free(pin);
+}
+
+static void test_wrap(const uint8_t *data, size_t size)
+{
+	CK_SESSION_HANDLE    session;
+	uint8_t              login_type = CKU_USER;
+	char                *pin = NULL;
+	CK_BYTE              pWrappedKey[4096];
+	CK_ULONG             pulWrappedKeyLen = sizeof(pWrappedKey);
+	CK_MECHANISM         mech = {0, NULL_PTR, 0};
+	CK_OBJECT_HANDLE     hWrappingKey;
+	CK_OBJECT_HANDLE     hkey;
+	const unsigned char *hkey_id;
+	const unsigned char *opt_id;
+	size_t               opt_id_len = 0, hkey_id_len = 0;
+
+	/* Set options */
+	if (set_mechanism(&data, &size, &mech) || size < 3)
+		return;
+	login_type = data[0];
+	data++; size--;
+	if (!(pin = extract_word(&data, &size)))
+		return;
+	opt_id = data;
+	opt_id_len = get_buffer(&opt_id, size, &data, &size, 256);
+	hkey_id = data;
+	hkey_id_len = get_buffer(&hkey_id, size, &data, &size, 256);
+
+	/* Initialize */
+	if (fuzz_pkcs11_initialize(data, size, NULL, &session) != CKR_OK)
+		goto end;
+
+	p11->C_Login(session, login_type, (CK_UTF8CHAR *) pin, strlen(pin));
+	if (!fuzz_find_object(session, CKO_SECRET_KEY, &hkey, hkey_id_len ? hkey_id : NULL, hkey_id_len))
+		goto fin;
+	if (!fuzz_find_object(session, CKO_PUBLIC_KEY, &hWrappingKey, opt_id_len ? opt_id : NULL, opt_id_len))
+		if (!fuzz_find_object(session, CKO_SECRET_KEY, &hWrappingKey, opt_id_len ? opt_id : NULL, opt_id_len))
+			goto fin;
+	p11->C_WrapKey(session, &mech, hWrappingKey, hkey, pWrappedKey, &pulWrappedKeyLen);
+
+fin:
+	p11->C_CloseSession(session);
+	p11->C_Finalize(NULL);
+end:
+	free(pin);
+}
+
+#define FILL_ATTR(attr, typ, val, len) {(attr).type=(typ); (attr).pValue=(val); (attr).ulValueLen=len;}
```

Comment:
please, expand this macro to multiple lines to make it more readable.

Reply:
Ok